src/java/org/apache/cassandra/repair/messages/RepairMessage.java

```diff
@@ -51,48 +65,165 @@ protected RepairMessage(RepairJobDesc desc)
+            case SNAPSHOT_MSG: return RepairRetrySpec.Verb.SNAPSHOT;
+            case CLEANUP_MSG: return RepairRetrySpec.Verb.CLEANUP;
+            default: return null;
+        }
+    }
+
+    public static <T> void sendMessageWithRetries(SharedContext.MessagingContext ctx, RepairMessage request, Verb verb, InetAddressAndPort endpoint, RequestCallback<T> finalCallback)
+    {
+        // 0 means no retries
+        sendMessageWithRetries(ctx, backoff(ctx, verb), request, verb, endpoint, finalCallback, 0);
+    }
+
+    public static <T> void sendMessageWithRetries(SharedContext.MessagingContext ctx, RepairMessage request, Verb verb, InetAddressAndPort endpoint)
+    {
+        sendMessageWithRetries(ctx, backoff(ctx, verb), request, verb, endpoint, new RequestCallback<>()
         {
             @Override
             public void onResponse(Message<Object> msg)
             {
-                logger.info("[#{}] {} received by {}", request.desc.parentSessionId, verb, endpoint);
-                // todo: at some point we should make repair messages follow the normal path, actually using this
+                // TODO (now): define
+            }
+
+            @Override
+            public void onFailure(InetAddressAndPort from, RequestFailureReason failureReason)
+            {
+                // TODO (now): define
+            }
+        }, 0);
+    }
+
+    private static <T> void sendMessageWithRetries(SharedContext.MessagingContext ctx, Backoff backoff, RepairMessage request, Verb verb, InetAddressAndPort endpoint, RequestCallback<T> finalCallback, int attempt)
+    {
+        // TODO (now, api): expose a way for callers to stop early even if backoff allows more progress
+        RequestCallback<T> callback = new RequestCallback<>()
+        {
+            @Override
+            public void onResponse(Message<T> msg)
+            {
+                finalCallback.onResponse(msg);
+            }
+
+            @Override
+            public void onFailure(InetAddressAndPort from, RequestFailureReason failureReason)
+            {
+                ErrorHandling allowed = errorHandlingSupported(ctx, endpoint, verb, request.parentRepairSession());
+                switch (allowed)
+                {
+                    case NONE:
+                        logger.error("[#{}] {} failed on {}: {}", request.parentRepairSession(), verb, from, failureReason);
+                        return;
+                    case TIMEOUT:
+                        finalCallback.onFailure(from, failureReason);
+                        return;
+                    case RETRY:
+                        int maxAttempts = backoff.maxAttempts();
+                        if ((maxAttempts < 0 || attempt < maxAttempts) && failureReason == RequestFailureReason.TIMEOUT)
+                        {
+                            ctx.optionalTasks().schedule(() -> sendMessageWithRetries(ctx, backoff, request, verb, endpoint, finalCallback, attempt + 1),
```

Comment:
Using the optional task pool doesn't seem like it would be too problematic, if all it's competing with are hint maintenance activities and other similar low-frequency things.

Reply:
I didn't want to create a scheduled pull for nothing... repair creates its own threads so I could make those scheduled... but the executor factory doesn't like scheduled pools that are of different sizes, so was kinda annoying.